lib/RestructureCFG/SelectScopePass.cpp

```diff
@@ -0,0 +1,261 @@
+//
+// Copyright rev.ng Labs Srl. See LICENSE.md for details.
+//
+
+#include "llvm/ADT/DepthFirstIterator.h"
+#include "llvm/ADT/PostOrderIterator.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/SetVector.h"
+#include "llvm/ADT/SmallPtrSet.h"
+#include "llvm/IR/BasicBlock.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/LLVMContext.h"
+#include "llvm/IR/Module.h"
+#include "llvm/Support/GenericDomTree.h"
+
+#include "revng/ADT/ReversePostOrderTraversal.h"
+#include "revng/RestructureCFG/ScopeGraphGraphTraits.h"
+#include "revng/RestructureCFG/ScopeGraphUtils.h"
+#include "revng/RestructureCFG/SelectScopePass.h"
+#include "revng/Support/Debug.h"
+#include "revng/Support/GraphAlgorithms.h"
+#include "revng/Support/IRHelpers.h"
+
+using namespace llvm;
+
+// Debug logger
+Logger<> SelectScopePassLogger("select-scope");
+
+static SetVector<BasicBlock *> getConditionalSuccessors(BasicBlock *N) {
+  // We employ a `SetVector` so that we do not take into account
+  // multiplicity for edges out of a conditional
+  SetVector<BasicBlock *> ConditionalSuccessors;
+  for (BasicBlock *Successor : children<Scope<BasicBlock *>>(N)) {
+    ConditionalSuccessors.insert(Successor);
+  }
+
+  return ConditionalSuccessors;
+}
+
+static SetVector<BasicBlock *> getPredecessors(BasicBlock *N) {
+  // It is important that we use a `SetVector` here in order to
+  // deduplicate the successors outputted by the `llvm::children` range
+  // iterator
+  SetVector<BasicBlock *> Predecessors;
+  for (auto *Predecessor : children<Inverse<Scope<BasicBlock *>>>(N)) {
+    revng_log(SelectScopePassLogger,
+              "  Predecessor: " + Predecessor->getName().str());
+    Predecessors.insert(Predecessor);
+  }
+
+  return Predecessors;
+}
+
+static SmallVector<BasicBlock *> getNodesToProcess(BasicBlock *N,
+                                                   BasicBlock *PostDominator) {
+  // We exploit the `Visited` set, by passing it to
+  // `ReversePostOrderTraversalExt`, in order to stop the visit at the
+  // `PostDominator`
+  std::set<BasicBlock *> Visited;
+  Visited.insert(PostDominator);
+
+  // We collect all the nodes between the conditional `PONode` and its
```

Comment:
I guess `PONode` should be `N` here.

Reply:
And there are other occurrences in comments below

Reply:
Right, I changed the name from `PONode` because it was to obscure and forgot to update the comments. In addition, I went with the name `Candidate` instead of `N` also for the `static` function here https://github.com/revng/revng/blob/a3b1cbac550057e016f37c8f9eb201e8ab290b82/lib/RestructureCFG/SelectScopePass.cpp#L53. It is true that in principle the parameter could also not be a conditional, but this helper has a very specific semantic, and it is a `static` function only used here, so I think this helps a lot when reading with understanding that we are collecting node between the `Conditional` and the `PostDominator`.